Use explicit `__init__` for `LocalGeometry`

LiamPattinson opened the pull request:
Resolves (partly) https://github.com/pyro-kinetics/pyrokinetics/issues/336.

Replaces general `__init__` functions of the form `def __init__(self, *args, **kwargs):` with more explicit argument lists. Uses the same default arguments as are currently in use.

- `default_inputs()` functions moved to class variables `DEFAULT_INPUTS`, as this allows them to be used to set default values in the `__init__` function and from outside the class (they're used in some `GKInput` classes).
- Added error checking on arrays passed to these functions to ensure they have the correct shape and size.

---

TODO:

- [x] Update `from_global_eq` to be a `@classmethod`.
- [x] Update `from_local_geometry` to be a `@classmethod`.
- [x]  ~~(Guidance needed) Should `normalise` be called as standard by the base `__init__`? If so, do that.~~ (EDIT: `normalise` should return a new `LocalGeometry` instead)


Stretch goals:

- [x] `LocalGeometry` could be a `@dataclass` which takes `R`, `Z`, `theta` and `b_poloidal` grids as inputs and therefore represent a 'non-parameterised' local geometry. Subclasses would instead be initialised with their parameters, and would generate `R`, `Z`, `theta` and `b_poloidal` grids from there. This would resolve the problem wherein many functions defined on the base class fail due to missing data that is only generated by subclasses.
- [ ] Given the above, `LocalGeometry` would remove the need for the `FluxSurface` class, which could be removed from the project.
- [x] ~~Initialisation with default values could still be achieved by updating the `default` method to be a `@classmethod`.~~ (EDIT: Not clear how this would work on the base class, probably better to avoid default behaviour where possible)
- [x] Update docstrings throughout `LocalGeometry` and subclasses

Commit:
Send lower case stella to upper case STELLA in appropriate instances. Note that normalisation conventions appear to use lower cases, and there is an input parameter to stella which uses the string "stella" in lower case.

Commit:
Try Except block for ky variable in case numerics.ky is not an array.

LiamPattinson committed:
Update publishing workflow

bpatel2107 committed:
Correct amplitude when loading eigenfunctions and add tests

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Flaake8 fix

bpatel2107 committed:
Merge remote-tracking branch 'origin/bugfix/eigenfunction_amp' into bugfix/eigenfunction_amp

LiamPattinson committed:
Use v4 artifact actions

Commit:
Add GKOutputReaderSTELLA to the __init__.py script.

Commit:
Initial changes to the output reader for stella to try to get basic output reading functionality (currently broken).

Commit:
Add a cheap example linear run for the stella gyrokinetic code.

mrhardman committed:
[skip ci] Apply black changes

bpatel2107 committed:
Merge pull request #339 from pyro-kinetics/bugfix/gene_dpdx_pm_sign

Bugfix/gene dpdx pm sign

bpatel2107 committed:
Merge branch 'unstable' into enhancements/gene_mxh

bpatel2107 committed:
Merge branch 'unstable' into stella_new

bpatel2107 committed:
Update test_roundtrip.py put test in right place

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Merge pull request #329 from pyro-kinetics/enhancements/gene_mxh

Add MXH to GENE

bpatel2107 committed:
Update Stella output reader and add into test suite

bpatel2107 committed:
Merge branch 'unstable' into stella_new

# Conflicts:
#	tests/test_roundtrip.py

bpatel2107 committed:
[skip ci] Apply black changes

Commit:
update stella.out.nc to have units for fluxes explicitly given

Commit:
Merge branch 'stella_new' of https://github.com/pyro-kinetics/pyrokinetics into stella_new

bpatel2107 committed:
Definition of inverse_lp was reverted accidentally when adding GKW. Test added to check it is correct using beta_star_scale in CGYRO

Commit:
Add comments.

Commit:
Add comments.

mrhardman committed:
[skip ci] Apply black changes

Commit:
A basic working version of the GKOutputReader for stella that is able to read flux (contributions) as a function of kx ky, with a fallback to the total flux if these variables are not available.

Commit:
Merge branch 'stella_new' of https://github.com/pyro-kinetics/pyrokinetics into stella_new

mrhardman committed:
[skip ci] Apply black changes

bpatel2107 committed:
Add in reader for nrg in h5 format

bpatel2107 committed:
[skip ci] Apply black changes

Commit:
Add a few collision options to the template file

Commit:
Handle the logical flags for correct conversion off collisions.

Commit:
Merge branch 'stella_new' of https://github.com/pyro-kinetics/pyrokinetics into stella_new

mrhardman committed:
[skip ci] Apply black changes

ZedThree committed:
Unpin version of scipy

Causes problems with Python 3.12. Hopefully original problem is
resolved now!

ZedThree committed:
CI: Test on Python 3.12

ZedThree committed:
CI: Bump checkout and setup-python actions

bpatel2107 committed:
Merge pull request #342 from pyro-kinetics/bugfix/inverse_lp_local_species

Definition of inverse_lp was reverted accidentally when adding GKW. T…

bpatel2107 committed:
Update test and add golden answer for stella

bpatel2107 committed:
flake8 fix

bpatel2107 committed:
Check for writing MXH in stella JETTO

bpatel2107 committed:
Merge pull request #338 from pyro-kinetics/stella_new

GK code stella input file conversion feature

bpatel2107 committed:
Update bunit calculation to use analytic solution

bpatel2107 committed:
Use default n0 for GS2

bpatel2107 committed:
Fix range location

bpatel2107 committed:
Merge branch 'enhancements/gs2_handle_n0' of github.com:pyro-kinetics/pyrokinetics into enhancements/gs2_handle_n0

bpatel2107 committed:
docstring typo

bpatel2107 committed:
Use scipy quad for exact evaluation of integrals

bpatel2107 committed:
Drop units in directory name when making Pyroscan with pint Quantity

bpatel2107 committed:
Merge branch 'bugfix/bunit_calculation' into enhancements/gs2_handle_n0

# Conflicts:
#	src/pyrokinetics/local_geometry/local_geometry.py

bpatel2107 committed:
Merge pull request #347 from pyro-kinetics/bugfix/pyroscan_directory_name

Drop units in directory name when making Pyroscan with pint Quantity

bpatel2107 committed:
Update definition of ky and tweak setters

bpatel2107 committed:
Merge branch 'enhancements/gs2_handle_n0' of github.com:pyro-kinetics/pyrokinetics into enhancements/gs2_handle_n0

bpatel2107 committed:
Fix ky in gs2 box param

bpatel2107 committed:
Merge pull request #346 from pyro-kinetics/bugfix/bunit_calculation

Update bunit calculation to use analytic solution

bpatel2107 committed:
Correct check on length of time array gene h5

bpatel2107 committed:
Merge branch 'refs/heads/unstable' into enhancements/gs2_handle_n0

bpatel2107 committed:
Merge pull request #343 from pyro-kinetics/feature/gene_nrg_h5

Add in reader for nrg in h5 format

bpatel2107 committed:
Multiply by Bref/B0 actual value rather than the unit

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Merge pull request #340 from pyro-kinetics/feature/update_publishing_workflow

Update publishing workflow

bpatel2107 committed:
Merge branch 'refs/heads/unstable' into bugfix/eigenfunction_amp

bpatel2107 committed:
Merge pull request #341 from pyro-kinetics/bugfix/eigenfunction_amp

Bugfix/eigenfunction amp

bpatel2107 committed:
Merge pull request #324 from pyro-kinetics/enhancements/gs2_handle_n0

Set up reading ky from n0 in gs2 input file

bpatel2107 committed:
Add correct mass factor in rhoref scaling

bpatel2107 committed:
Merge pull request #348 from pyro-kinetics/bugfix/rhoref_mass_scaling

Add correct mass factor in rhoref scaling

bpatel2107 committed:
Update README.md

bpatel2107 committed:
Merge pull request #350 from pyro-kinetics/docs/update_readme_python_version

Update README.md

bpatel2107 committed:
Merge branch 'unstable' into enhancements/syn_hk_dbs

bpatel2107 committed:
Merge pull request #294 from pyro-kinetics/enhancements/syn_hk_dbs

wip: highk/dbs synthetic diagnostic

juanruizruiz committed:
synthetic hk dbs doc

bpatel2107 committed:
Use correct variable for fast ion nucleon

bpatel2107 committed:
Merge pull request #355 from pyro-kinetics/bugfix/pfile_fast_ions

Use correct variable for fast ion nucleon

LiamPattinson committed:
Fix geometry unit conversions using simulation units

LiamPattinson committed:
Fix local species normalisation with simulation units

LiamPattinson committed:
Minor edits to units docs, local geometry units test

bpatel2107 committed:
Merge pull request #357 from pyro-kinetics/bugfix/normalise_local_geometry

Bugfix: unit conversions using simulation units

LiamPattinson committed:
Merge branch 'unstable' into unpin-scipy

LiamPattinson committed:
Wrap scipy trapezoid functions

Includes various other minor fixes:
- test_imas.py global-level code moved into fixtures, no longer
  writes a local file pyro_ids.h5 whenever the tests are run
- replaced a few 'assert np.allclose' with
  'np.testing.assert_allclose'
- Fixed scipy.integrate.simpson warnings about named keyword args

LiamPattinson committed:
[skip ci] Apply black changes

LiamPattinson committed:
Pin idspy-dictionaries version

bpatel2107 committed:
Merge pull request #358 from pyro-kinetics/bugfix/incompatible_idspy_dictionaries

Bugfix: Pin `idspy-dictionaries` version

LiamPattinson committed:
Fix 'UnitStripped' warnings, fix broken Fourier/MXH tests

LiamPattinson committed:
Merge branch 'unstable' into unpin-scipy

LiamPattinson committed:
Fix unused import

LiamPattinson committed:
Fix golden answer test issues

LiamPattinson committed:
Install setuptools with Python 3.12 for jetto-tools compatibility

LiamPattinson committed:
Change entry_points behaviour for Python 3.10+

bpatel2107 committed:
Read trpeps from geometry output file if it exists

bpatel2107 committed:
black formatting

LiamPattinson committed:
Fix pint.UnitStrippedWarning throughout

bpatel2107 committed:
Add original_filename to GKInputGENE when reading output

bpatel2107 committed:
Look for _0000 type files

bpatel2107 committed:
Use Path object when splitting filename

bpatel2107 committed:
Pull out tpreps in a functions

bpatel2107 committed:
Make major radius initial guess for Lref in set_ref_values and take magnitude of conversion factor in normalisation.py

bpatel2107 committed:
black files

bpatel2107 committed:
Merge pull request #359 from pyro-kinetics/feature/gene_trpeps_geometry_file

Feature/gene trpeps geometry file

bpatel2107 committed:
Merge pull request #345 from pyro-kinetics/unpin-scipy

Unpin scipy

bpatel2107 committed:
Add units to eigenfunctions when loaded manually

bpatel2107 committed:
Add test for eigenfunction units

bpatel2107 committed:
Add check for box_size when no kx is set with a test.

bpatel2107 committed:
Merge pull request #361 from pyro-kinetics/bugfix/eigenfunction_units

Add units to eigenfunctions when loaded manually

bpatel2107 committed:
Add method to enforce quasineutrality in LocalSpecies

hgd511 committed:
Merge pull request #362 from pyro-kinetics/bugfix/nl_grid_no_kx

Add check for box_size when no kx is set with a test.

bpatel2107 committed:
Merge pull request #365 from pyro-kinetics/feature/enforce_quasineutrality_local_species

Add method to enforce quasineutrality in LocalSpecies

bpatel2107 committed:
Remove extra species namelist from inputs when removed

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Add in FourierGENE to gene.py and tweak normalisation defn of aspect ratio

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Remove unnecessary to_base_units

bpatel2107 committed:
Merge pull request #366 from pyro-kinetics/bugfix/remove_gkinput_species

Remove extra species namelist from inputs when removed

bpatel2107 committed:
Set default norm convention when reading IDS GKOutput and add test

hgd511 committed:
Merge pull request #371 from pyro-kinetics/bugfix/default_convention_ids

Set default norm convention when reading IDS GKOutput and add test

LiamPattinson committed:
Replace np.asfarray(x) with np.asarray(x, dtype=float)

LiamPattinson committed:
Use 64 bit floats in CGYRO/GKW

bpatel2107 committed:
Merge pull request #370 from pyro-kinetics/enhancements/gene_fourier_reading

Enhancements/gene fourier reading

bpatel2107 committed:
Switch TGLF to use MillerTurnbull

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Correct miller_turnbull value

bpatel2107 committed:
Correct s_delta in tglf

bpatel2107 committed:
Merge pull request #373 from pyro-kinetics/bugfix/tglf_miller_turnbull

Switch TGLF to use MillerTurnbull

bpatel2107 committed:
Update to loop over all wavenumber/eigenmode and get fluxes/field to return dimensions directly

bpatel2107 committed:
Merge branch 'unstable' into enhancements/ids_multiple_mode_wavenumbers

bpatel2107 committed:
Set growth rate tolerance to 0.0 if no time dimension in eigenvalues

bpatel2107 committed:
Add TGLF to test and correct mode indexing

bpatel2107 committed:
Fix TGLF mode indexing for linear

bpatel2107 committed:
Add TGLF linear/nonlinear into test and select fields/fluxes as available

bpatel2107 committed:
Update test to account for 0 indexing of TGLF mode number

bpatel2107 committed:
Add default in ky from tglf get_numerics

bpatel2107 committed:
Merge pull request #374 from pyro-kinetics/enhancements/ids_multiple_mode_wavenumbers

Update to loop over all wavenumber/eigenmode and get fluxes/field to …

juanruizruiz committed:
adding documentation for synthetic hk-dbs

juanruizruiz committed:
Apply suggestions from code review

Co-authored-by: Bhavin Patel <15210802+bpatel2107@users.noreply.github.com>

bpatel2107 committed:
Update syn_hkdbs.rst

juanruizruiz committed:
iteration of syn diag documentation

bpatel2107 committed:
Merge branch 'unstable' into docs/synthetic_hk_dbs

bpatel2107 committed:
Update syn_hkdbs.rst

bpatel2107 committed:
Update syn_hkdbs.rst

bpatel2107 committed:
Merge pull request #351 from pyro-kinetics/docs/synthetic_hk_dbs

Adding documentation for synthetic highk/dbs

bpatel2107 committed:
Merge pull request #375 from pyro-kinetics/bugfix/default_ky_tglf

Add default in ky from tglf get_numerics

bpatel2107 committed:
Use Coords object to define fields coordinate

bpatel2107 committed:
Add phi as True to Stella field coord

bpatel2107 committed:
Add check to IMAS if NL field is there or not

bpatel2107 committed:
Merge pull request #379 from pyro-kinetics/bugfix/gk_output_field_coord

Bugfix/gk output field coord

bpatel2107 committed:
Add check for new position of rhostar in cgyro output

bpatel2107 committed:
Merge pull request #380 from pyro-kinetics/bugfix/cgyro_rhostar

Add check for new position of rhostar in cgyro output

bpatel2107 committed:
Add default for omega0 in GENE

bpatel2107 committed:
Check for field before adding and change confusing variable name

bpatel2107 committed:
Use correct if statement when checking for fields

bpatel2107 committed:
Careful selection of which fluxes_* to fill/load in IDS depending on available data

bpatel2107 committed:
Try to load fluxes as a function of time first

bpatel2107 committed:
Change test to handle new default dimension of fluxes

bpatel2107 committed:
Add reader R, Z and B_pol from magn_geometry file. Also fix for 1 species GENE sims

bpatel2107 committed:
Merge pull request #382 from pyro-kinetics/bugfix/gene_omega0_default

GENE IMAS fixes

bpatel2107 committed:
Apply dz/dzprime correction when edge_opt = True in GENE

bpatel2107 committed:
Add IMAS h5 reader in for Kinetics/Equilibrium objects

bpatel2107 committed:
Add in basic tests for IMAS readerand pass Equilibrium object to Kinetics when needed

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Update library requirements and README/intro page on the docs

bpatel2107 committed:
Update pfile tests with new eq input

LiamPattinson committed:
Merge branch 'unstable' into bugfix/numpy_2_compatibility

bpatel2107 committed:
Temporarily pin numpy~1.24

LiamPattinson committed:
Merge pull request #372 from pyro-kinetics/bugfix/numpy_2_compatibility

Numpy 2.0 Compatibility

bpatel2107 committed:
Merge branch 'refs/heads/unstable' into enhancements/ids_equ_kinetics_reader

bpatel2107 committed:
Unpin numpy~1.24

bpatel2107 committed:
Transport psi grid and use better psi_boundary value

bpatel2107 committed:
Add rotation and remove units when plotting

bpatel2107 committed:
Add golden answer tests for IMAS

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Update src/pyrokinetics/kinetics/imas.py

Co-authored-by: Peter Hill <peter.hill@york.ac.uk>

bpatel2107 committed:
Apply suggestions from code review

Co-authored-by: Peter Hill <peter.hill@york.ac.uk>

bpatel2107 committed:
Handle reviewer comments

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Remove duplicate import matplotlib

bpatel2107 committed:
Add matplotlib back for type checking

bpatel2107 committed:
Add annotation import

bpatel2107 committed:
Merge pull request #383 from pyro-kinetics/enhancements/ids_equ_kinetics_reader

IMAS Equilibrium/Kinetics Reader

LiamPattinson committed:
Document LocalGeometry vars with type checking

LiamPattinson committed:
Document LocalGeometry vars with type checking

LiamPattinson committed:
Update LocalGeometry variable docs

LiamPattinson committed:
Update LocalGeometry variable docs

LiamPattinson committed:
Replace LocalGeometry default input functions with ClassVars

LiamPattinson committed:
Replace LocalGeometry default input functions with ClassVars

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometry

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometry

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMiller

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMiller

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMillerTurnbull

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMillerTurnbull

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMXH

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryMXH

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryFourierCGYRO

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryFourierCGYRO

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryFourierGENE

LiamPattinson committed:
[TESTS FAIL] Add explicit __init__ to LocalGeometryFourierGENE

LiamPattinson committed:
[TESTS FAIL] Fix missing optional args in MXH

LiamPattinson committed:
[TESTS FAIL] Fix missing optional args in MXH

LiamPattinson committed:
[TESTS FAIL] LocalGeometry.from_gk_data uses new __init__

LiamPattinson committed:
[TESTS FAIL] LocalGeometry.from_gk_data uses new __init__

LiamPattinson committed:
Fix failing normalisation tests

LiamPattinson committed:
Fix failing normalisation tests

LiamPattinson committed:
[skip ci] Apply black changes

LiamPattinson committed:
[skip ci] Apply black changes

LiamPattinson committed:
Remove walrus operator trick to keep Python 3.9 happy

LiamPattinson committed:
Remove walrus operator trick to keep Python 3.9 happy

LiamPattinson committed:
Use TypeAlias from typing_extensions to keep Python 3.9 happy

LiamPattinson committed:
Use TypeAlias from typing_extensions to keep Python 3.9 happy

LiamPattinson committed:
Update LocalGeometry.from_global_eq to classmethod

LiamPattinson committed:
Update LocalGeometry.from_global_eq to classmethod

LiamPattinson commented:
The method `from_global_eq` is now a `@classmethod`, meaning it can be used as an alternative constructor in place of the basic `__init__`, and can no longer be used to modify an existing `LocalGeometry`. 

> [!Note]
> The ability to set `pyro.local_geometry` with a string has been removed. Rather than converting the currently loaded `LocalGeometry` to the desired type, this would instead replace it with an empty object, which risks allowing uninitialised data to enter the program.
> 
> This method was only used in the function `load_local_geometry` to pre-emptively set the type of `LocalGeometry` before updating it with `from_global_eq`. With `from_global_eq` being a classmethod, this is no longer needed.
>
> This method was also used in some examples scripts, but they were easily modified to use a safer method.
>
> (This also allowed me to delete a 2 year old FIXME comment I'd left in `Pyro`, which I'm quite happy about)

--- 

Also fixed a minor error in the `Equilibrium` docs while I was at it

LiamPattinson committed:
Update LocalGeometry.from_local_geometry to classmethod

LiamPattinson committed:
Update LocalGeometry.from_local_geometry to classmethod

LiamPattinson commented:
The latest commit changes `from_local_geometry` to be a `@classmethod`:

```python
# Before
local_geometry = LocalGeometryMXH() # Empty object
local_geometry.from_local_geometry(other_local_geometry)

# After
local_geometry = LocalGeometryMXH.from_local_geometry(other_local_geometry)
```

---

From here, I could either carry on working on this to complete the additional goals listed above, or (if you're happy with the changes) we could merge this and I'll raise a new PR once I've made some progress. I'm hoping to be done with the further goals roughly by the end of next week.

bpatel2107 commented on src/pyrokinetics/local_geometry/local_geometry.py:
```diff
@@ -1,35 +1,138 @@
         "a_minor": 1.0,
         "Fpsi": 0.0,
-        "B0": None,
+        "B0": 0.0,
```
I'm nervous about setting this to 0 as we sometimes to `1/B0`...

ZedThree replied:
In some ways it would make more sense to _not_ have a default value at all for basically all of these

LiamPattinson replied:
Good point, but the current default value of `None` is also a problem in that case (albeit one that will throw an error immediately rather than letting `inf`/`NaN` propagate throughout the code).

Is there a more sensible default value for `B0`? If the current defaults are pulled from a standard test case, is `B0` defined there?

LiamPattinson replied:
@ZedThree I'd like to eventually get these classes to a point where none of the inputs are optional (besides things like `delta` in `LocalGeomtryMXH`), but the option to build one with sensible defaults could still be achieved with a class method, e.g. `LocalGeometry.default()`, as that would make testing a lot easier.

bpatel2107 replied:
I think we usually have checks for `B0` usually like 

```python
if local_geometry.B0:
    beta = 1 / local_geometry.B0**2
else:
    beta = 0.0
```

Hence the default of `None` as having `beta=0.0` by default is safest.

bpatel2107 commented:
This looks great, makes `LocalGeometry` more like the other classes and I think being consistent overall is a good idea. Making it a `@dataclass` seems likea good way to go ahead.

`normalise` adds/converts the units of the different `LocalGeometry` parameters. When all is said and done they should have units. The issue becomes whether or not the units or way to convert units is known when initialising the object. When loading from `Equilibrium` it is straightforward, but when loading from input files there may be times where the units are not immediately obvious without looking at other parts of the input file. This could be handled in the specific code such that `from_gk_data` includes the appropriate units/convention (should be the called `load_from_gk_data`?). So `normalise` should be in the `__init__` but we need to make sure we know the appropriate units.

LiamPattinson committed:
Fix for FreeQDSK 0.5 breaking changes

LiamPattinson committed:
Pin FreeQDSK version to 0.5.x

LiamPattinson committed:
Remove unnecessary G-EQDSK file checks

bpatel2107 committed:
Merge pull request #387 from pyro-kinetics/bugfix/freeqdsk_0.5

Fix for FreeQDSK 0.5 breaking changes

LiamPattinson committed:
Merge branch 'unstable' into feature/local_geometry_initialisers

LiamPattinson committed:
Merge branch 'unstable' into feature/local_geometry_initialisers

LiamPattinson committed:
Add fit_params function to LocalGeometry, remove repeat code

Function wraps up parameter fitting logic and error checking
from _set_shape_coeffecients.

LiamPattinson committed:
Add fit_params function to LocalGeometry, remove repeat code

Function wraps up parameter fitting logic and error checking
from _set_shape_coeffecients.

LiamPattinson committed:
Rename test_fourier.py to test_fourier_gene.py

LiamPattinson committed:
Rename test_fourier.py to test_fourier_gene.py

LiamPattinson committed:
Remove b_poloidal_even_space

LiamPattinson committed:
Remove b_poloidal_even_space

LiamPattinson committed:
Use b_poloidal passed to _set_shape_coefficients

Miller geometries were previously bypassing this and
using self.b_poloidal_eq instead

LiamPattinson committed:
Use b_poloidal passed to _set_shape_coefficients

Miller geometries were previously bypassing this and
using self.b_poloidal_eq instead

LiamPattinson committed:
Use local theta in Miller fit_params

Was previously using self.theta_eq

LiamPattinson committed:
Use local theta in Miller fit_params

Was previously using self.theta_eq

LiamPattinson committed:
Remove R_eq, Z_eq, theta_eq, b_poloidal_eq

Simplifies design to use just one variant of the grids throughout

LiamPattinson committed:
Remove R_eq, Z_eq, theta_eq, b_poloidal_eq

Simplifies design to use just one variant of the grids throughout

LiamPattinson committed:
Only set shape coefficients in _set_shape_coefficients

LiamPattinson committed:
Only set shape coefficients in _set_shape_coefficients

LiamPattinson committed:
Bundle all shape coefficients into a single NamedTuple

Aiming to avoid all references to 'self' during parameter fitting,
so need a convenient way to pass all shape parameters down the
stack.

New ShapeParams uses a decorator to add some utility funcitons,
including the ability to unpack and repack when passing data
into/out of the least squares call.

LiamPattinson committed:
Bundle all shape coefficients into a single NamedTuple

Aiming to avoid all references to 'self' during parameter fitting,
so need a convenient way to pass all shape parameters down the
stack.

New ShapeParams uses a decorator to add some utility funcitons,
including the ability to unpack and repack when passing data
into/out of the least squares call.

LiamPattinson committed:
[TESTS FAIL] Miller fitting uses only classmethods

LiamPattinson committed:
[TESTS FAIL] Miller fitting uses only classmethods

bpatel2107 committed:
Add correct defn of Bpoloidal from GENE, set up lref=magnetic_axis and add tests in.

bpatel2107 committed:
Merge branch 'refs/heads/unstable' into enhancements/gene_magn_geometry_reader

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Add in drhotor_dr term to gradients

bpatel2107 committed:
Add default value for magnetic_axis_radius everywhere

bpatel2107 committed:
Correct few defn and tests

bpatel2107 committed:
Fix dpsidr units and update tests

bpatel2107 committed:
Correct angle calculation when Z0 not in data

bpatel2107 committed:
Fix floating point error in fourier_gene.py

bpatel2107 committed:
black fix

LiamPattinson committed:
[TESTS FAIL] Miller turnbull fitting only uses classmethods

LiamPattinson committed:
[TESTS FAIL] Miller turnbull fitting only uses classmethods

bpatel2107 committed:
Fixed bug in Unit conversion when source and destination powers dont match, made func to get local_species names and and in n_moments kwarg to MXH

LiamPattinson committed:
[TESTS FAIL] MXH fitting uses only classmethods

LiamPattinson committed:
[TESTS FAIL] MXH fitting uses only classmethods

LiamPattinson committed:
[TESTS FAIL] Fourier GENE fitting only uses classmethods

LiamPattinson committed:
[TESTS FAIL] Fourier GENE fitting only uses classmethods

LiamPattinson committed:
Fourier CGYRO fitting uses only classmethods

LiamPattinson committed:
Fourier CGYRO fitting uses only classmethods

LiamPattinson committed:
Add Derivatives NamedTuple

Cleans up a long return type and ensures derivatives aren't
returned in the wrong order. However, doesn't stop the tuple
from being unpacked in the wrong order!

LiamPattinson committed:
Add Derivatives NamedTuple

Cleans up a long return type and ensures derivatives aren't
returned in the wrong order. However, doesn't stop the tuple
from being unpacked in the wrong order!

LiamPattinson committed:
Move Float and Array type aliases to units.py

LiamPattinson committed:
Move Float and Array type aliases to units.py

LiamPattinson committed:
_set_shape_coefficeints renamed _fit_shape_params, now returns ShapeParams

Also changed to classmethod

LiamPattinson committed:
_set_shape_coefficeints renamed _fit_shape_params, now returns ShapeParams

Also changed to classmethod

LiamPattinson committed:
Calculate params before calling init in from_global_eq/from_local_geometry

A step closer to having the base class __init__ be a dataclass,
and from_global_eq/from_local_geometry to determine all inputs.

from_gk_data should be replaced by subclass-specific __init__'s

Fixes some units headaches in MXH

LiamPattinson committed:
Calculate params before calling init in from_global_eq/from_local_geometry

A step closer to having the base class __init__ be a dataclass,
and from_global_eq/from_local_geometry to determine all inputs.

from_gk_data should be replaced by subclass-specific __init__'s

Fixes some units headaches in MXH

LiamPattinson committed:
[TESTS FAIL] Rearrange from_gk_data

Fails when delta/s_delta are passed to create MXH, as it can't handle
the redundant values.

Planning to replace this with a non-classmethod utility function
which will be called by each subclass. Calls to from_gk_data will
be replaced with simple __init__ calls, and MXH.__init__ can
handle the edge case individually.

LiamPattinson committed:
[TESTS FAIL] Rearrange from_gk_data

Fails when delta/s_delta are passed to create MXH, as it can't handle
the redundant values.

Planning to replace this with a non-classmethod utility function
which will be called by each subclass. Calls to from_gk_data will
be replaced with simple __init__ calls, and MXH.__init__ can
handle the edge case individually.

LiamPattinson committed:
Replace from_gk_data with _init_with_shape_params

This is used by from_global_eq and from_local_geometry, as well
as all subclass __init__ functions. Provides a single unified
method for setting up a LocalGeometry after fitting shape
parameters.

Problem: Had to introduce variable overwrite_dpsidr, as previous
behaviour was to accept dpsidr when provided from an equilibrium
or a local geometry, but to always overwrite it when reading from
a GK input file. This feels like a bug, and the correct behaviour
should be to do this only if dpsidr is not provided, but it the
behaviour has been maintained in order to keep the tests correct.

LiamPattinson committed:
Replace from_gk_data with _init_with_shape_params

This is used by from_global_eq and from_local_geometry, as well
as all subclass __init__ functions. Provides a single unified
method for setting up a LocalGeometry after fitting shape
parameters.

Problem: Had to introduce variable overwrite_dpsidr, as previous
behaviour was to accept dpsidr when provided from an equilibrium
or a local geometry, but to always overwrite it when reading from
a GK input file. This feels like a bug, and the correct behaviour
should be to do this only if dpsidr is not provided, but it the
behaviour has been maintained in order to keep the tests correct.

bpatel2107 committed:
Merge pull request #388 from pyro-kinetics/enhancements/gene_magn_geometry_reader

Enhancements/gene magn geometry reader

LiamPattinson committed:
LocalGeomtry is a dataclass

Now takes R, Z, theta, B_poloidal etc as input args

LiamPattinson committed:
LocalGeomtry is a dataclass

Now takes R, Z, theta, B_poloidal etc as input args

LiamPattinson committed:
Clean up redundant functions

LiamPattinson committed:
Clean up redundant functions

LiamPattinson committed:
Refactor ShapeParams to use dataclass instead of NamedTuple

Allows use of inheritance instead of a decorator, and makes
type hinting much easier.

LiamPattinson committed:
Refactor ShapeParams to use dataclass instead of NamedTuple

Allows use of inheritance instead of a decorator, and makes
type hinting much easier.

LiamPattinson committed:
Ensure verbose and max_cost are passed through fitting routine

LiamPattinson committed:
Ensure verbose and max_cost are passed through fitting routine

LiamPattinson commented on src/pyrokinetics/local_geometry/local_geometry.py:
```diff
@@ -1,152 +1,341 @@
+        R, Z = self._flux_surface(theta, Rmaj, Z0, rho, params)
+        derivatives = self._RZ_derivatives(theta, rho, params)
+        bunit_over_b0 = self._bunit_over_b0(Rmaj, Z0, rho, params)
+        if overwrite_dpsidr:
+            dpsidr = rho * bunit_over_b0 / q
```
When initialising a local geometry, the code was previously overwriting `dpsidr` if it was provided via a GK input file, but not if it was determined via an equilibrium or another local geometry. This feels like a bug.

This check is included as otherwise a few tests fail that I can't verify for myself. It also necessitates an `overwrite_dpsidr` variable in all of the subclass `__init__` functions. If this is unintended behaviour, I can remove it throughout.

bpatel2107 replied:
`dpsidr` should be determined from `LocalGeometry` parameters alone, you never actually specify it as an input to a code. You can of course get it from a full equilibrium so that could be a good check but in theory you should always take the one from the `LocalGeometry` parameters, its like a fit I guess.  I guess it should always be overwritten.

LiamPattinson replied:
I've been playing around with this and I ran into a bit of a units issue. If we build from a GK input file, then everything is just a `float` until we attach units during the normalisation step, after which `dpsidr` is assigned units of `lref * bref`. However, if we calculate it from an equilibrium file, `dpsidr` will have units of just `[length]`, as `rho * bunit_over_b0 / q` has units `[length] [dimensionless] / [dimensionless]`, and when we try to normalise this to `lref * bref` it throws an error. I think we need to use the provided `dpsidr` when building from an equilibrium, but otherwise should be fine to recalculate.

LiamPattinson replied:
I did some further playing around and it seems this issue appears when building from another local geometry too. It looks like the best solution is to use `dpsidr` if it's available, and only calculate it as a backup. I've come up with a cleaner solution to the problem though, so it's not a big deal -- not sure why I thought I needed `overwrite_dpsidr`, as it seems just using a default value of `None` is sufficient.

bpatel2107 replied:
The issue here with `bunit_over_b0` is that it technically is `Bunit` with units of `B0` so when `bref` = `B0` then it dimensionally makes sense. To be honest it may be worth adding in the correct units for `Bunit` which is a just a magnetic  field rather than having `Bunit/B0`.

LiamPattinson commented on tests/local_geometry/test_fourier_gene.py:
```diff
@@ -134,38 +121,43 @@ def test_default_bunit_over_b0(generate_miller):
-                + 2.0 * np.cos(0.523598775598299 * np.sin(theta))
-            ),
-        ),
+        # FIXME: L. Pattinson 2024-09-26
+        # This test is failing because the theta grid defined at the top of the function
+        # is different to fourier.theta (formerly fourier.theta_eq). The latter was set
+        # up using the theta calculation at the top of _set_shape_coefficients, and does
+        # not correspond to the original miller theta grid. I don't know why
+        # fourier.theta_eq was being set this way, nor why this test is comparing grad_r
+        # calculated at different angles. I also don't know how to regenerate the data
+        # using sympty.
+        # (
+        #     {"kappa": 2.0, "delta": 0.5, "s_kappa": 0.5, "s_delta": 0.2, "shift": 0.1},
+        #     lambda theta: 2.0
+        #     * np.sqrt(
+        #         0.25
+        #         * (0.523598775598299 * np.cos(theta) + 1) ** 2
+        #         * np.sin(theta + 0.523598775598299 * np.sin(theta)) ** 2
+        #         + np.cos(theta) ** 2
+        #     )
+        #     / (
+        #         2.0
+        #         * (0.585398163397448 * np.cos(theta) + 0.5)
+        #         * np.sin(theta)
+        #         * np.sin(theta + 0.523598775598299 * np.sin(theta))
+        #         + 0.2 * np.cos(theta)
+        #         + 2.0 * np.cos(0.523598775598299 * np.sin(theta))
+        #     ),
+        # ),
```
I had to disable this test and a few others for FourierGENE/CGYRO, as I think they were comparing values calculated on different theta grids, and I can't figure out how the test parameters were calculated. Where the tests say the data was determined using sympy, it would be good if the code used to generate the data could be included in the docstrings.

bpatel2107 replied:
These are analytic solutions for `grad_r` in a simplified geometry so they should always be true. We can look at getting the sympy version written up somewhere

bpatel2107 replied:
Example code using sympy to get the golden answers. Given that we fit say a FourierGENE to a Miller fit we expect it to be close but not exact. Maybe the tolerance should be increased though nothing here should change an answer that already worked... Sometimes we need to be careful with integrals used in minimising `b_poloidal` and how they are set up with boundaries in `theta` Maybe the `theta` grid has changed which could tweak the integral and modify the fit

```python
from sympy import Symbol, simplify, cos, sin, asin, sqrt


def simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta):

    theta = Symbol("theta")

    # Not used but nice to see
    R = Rmaj + rho * (cos(theta + asin(delta) * sin(theta)))
    Z = rho * kappa * sin(theta)

    dRdr = (
        shift
        + cos(theta + asin(delta) * sin(theta))
        - sin(theta + asin(delta) * sin(theta)) * sin(theta) * s_delta
    )

    dRdtheta = (
        -rho * sin(theta + asin(delta) * sin(theta)) * (1 + asin(delta) * cos(theta))
    )

    dZdr = kappa * sin(theta) + s_kappa * kappa * sin(theta)
    dZdtheta = kappa * rho * cos(theta)
    g_tt = dRdtheta**2 + dZdtheta**2

    grad_r = sqrt(g_tt) / (dRdr * dZdtheta - dRdtheta * dZdr)

    return simplify(grad_r)


# Simple circular plasma
rho = 0.5

Rmaj = 3.0
shift = 0.0

kappa = 1.0
s_kappa = 0.0

delta = 0.0
s_delta = 0.0

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for circular plasma: {my_grad_r}")


# Elongation gradient
rho = 0.5

Rmaj = 3.0
shift = 0.0

kappa = 1.0
s_kappa = 1.0

delta = 0.0
s_delta = 0.0

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for circular plasma with s_kappa: {my_grad_r}")


# High shaping
rho = 0.5

Rmaj = 3.0
shift = 0.1

kappa = 2.0
s_kappa = 1.0

delta = 0.5
s_delta = 0.2

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for shaped plasma: {my_grad_r}")

```

Running this code results in 
```
Grad_r for circular plasma: 1.00000000000000
Grad_r for circular plasma with s_kappa: 1.0/(sin(theta)**2 + 1)
Grad_r for shaped plasma: sqrt(0.25*(0.523598775598299*cos(theta) + 1.0)**2*sin(theta + 0.523598775598299*sin(theta))**2 + 1.0*cos(theta)**2)/(2.0*(0.523598775598299*cos(theta) + 1.0)*sin(theta)*sin(theta + 0.523598775598299*sin(theta)) + 1.0*(-0.2*sin(theta)*sin(theta + 0.523598775598299*sin(theta)) + cos(theta + 0.523598775598299*sin(theta)) + 0.1)*cos(theta))
```

LiamPattinson replied:
Thanks for this, I tried adapting the code to also generate `R` and `Z` to make sure we were comparing the same flux surfaces. It looks like we are, but each value of theta corresponds to a different point along the flux surface:

![sympy_vs_fourier_gene](https://github.com/user-attachments/assets/54628186-01b3-4962-853f-70dad6cef87e)

It's a bit too late on a Friday for me to figure out what's going on here. Is it the case that the definition of theta is different in the Miller geometry compared to the Fourier geometries?

bpatel2107 replied:
Yes the definition is different which makes a lot of this quite annoying. You need to determine theta when given an array of R and Z. You can see the definitions would be different when looking at the parameterisation for each type.

LiamPattinson replied:
I think I have this resolved in the latest commit, but it's a bit hacky! I tried to find an analytic way to convert between the two theta systems, but in the end gave up and just added a small routine that takes R, Z, and theta from one local geometry and finds which Miller-theta should correspond to that using a least squares fit.

LiamPattinson reviewed (commented):
Finally managed to get somewhere with this refactor, though it looks like I took so long that I now have a really messy merge on my hands...

`LocalGeometry` is now a dataclass, and local geometries are now always built using either their own `__init__`, `from_global_eq`, or `from_local_geometry`. They have a consistent set of attributes whichever method was used, and no further set up is required after initialisation (e.g. no need to assign `bunit_over_b0` after building the object).

The fitting process for shaping coefficients has been modified to use `@classmethod`'s and `@staticmethod`'s throughout. This means that `self` is never referenced during the fitting process, and all necessary data is determined before initialisation (see `LocalGeometry._init_with_shape_params`). I had to add some extra machinery to get this to work in a generic manner, such as the `ShapeParams` classes and `_shape_params` property. This reduces the amount of repeat code in subclasses, at the cost of a more complex base class.

There are still some tasks to complete, two of which I've listed in comments. Further work includes:

- [x] Clean up docs in subclasses
- [x] Remove defaults from subclass `__init__` functions. This may be challenging, as quite a lot of the code and the tests currently depend on default behaviour.
- [x] `from_global_eq` should not modify the normalisation -- the responsibility for this step should be handed to `Pyro`.
- [x] `normalise` should return a new local geometry

LiamPattinson committed:
Merge branch 'unstable' into feature/local_geometry_initialisers

LiamPattinson committed:
Merge branch 'unstable' into feature/local_geometry_initialisers

LiamPattinson commented on src/pyrokinetics/gk_code/gene.py:
```diff
@@ -416,62 +411,63 @@ def get_local_geometry(self) -> LocalGeometry:
+            # Create a mock LocalGeometry using geometry dict and fit to it.
+            # This will get a new set of shaping parameters, which are then modified.
+
+            mock = LocalGeometryFourierGENE.from_local_geometry(
+                LocalGeometry(
+                    psi_n=local_geometry.psi_n,
+                    rho=geometry_dict["rho"] * lref,
+                    a_minor=local_geometry.a_minor,
+                    Rmaj=geometry_dict["Rmaj"] * lref,
+                    Z0=geometry_dict["Z0"] * lref,
+                    Fpsi=local_geometry.Fpsi,
+                    FF_prime=local_geometry.FF_prime,
+                    B0=local_geometry.B0,
+                    bunit_over_b0=local_geometry.bunit_over_b0,
+                    dpsidr=geometry_dict["dpsidr"] * bref * lref,
+                    q=local_geometry.q,
+                    shat=geometry_dict["shat"],
+                    beta_prime=geometry_dict["beta_prime"] * bref**2 / lref,
+                    R=geometry_dict["R"] * lref,
+                    Z=geometry_dict["Z"] * lref,
+                    b_poloidal=geometry_dict["b_poloidal"] * bref,
+                    theta=geometry_dict["theta"],
+                    bt_ccw=local_geometry.bt_ccw,
+                    ip_ccw=local_geometry.ip_ccw,
+                    dRdtheta=local_geometry.dRdtheta,
+                    dRdr=local_geometry.dRdr,
+                    dZdtheta=local_geometry.dZdtheta,
+                    dZdr=local_geometry.dZdr,
+                )
```
I've finished merging and attempted to recreate the new GENE features using my new methods, but I can't tell if I've done so correctly or not. This section of the code looks like it's untested, and I don't really know what it's doing.

bpatel2107 replied:
Essentially the full data needed to populate `LocalGeometry` is contained inside a different file which requires a bit of post-processing to get the data in a form that can actually be read by `LocalGeometry`. What you do here looks to be correct. This is tested here https://github.com/pyro-kinetics/pyrokinetics/blob/765d0f1032470937926431847a9f87efefcdcd41/tests/local_geometry/test_fourier.py#L383

LiamPattinson replied:
That test doesn't seem to touch the new code in `gk_code/gene.py`, so I'm not fully confident that my changes haven't broken something. I can leave it alone for now since it's slightly outside the scope of the PR, but it would be good to get a test in that uses one of those new geometry files alongside a GENE input file directly.

bpatel2107 replied:
Right I see what you mean! I just fit a `FourierGENE` to the GEQDSK and compare to a new geometry file I made independently. I can add that in but yes its outside the scope of this PR

LiamPattinson committed:
Clean up and comments on GENE tracer_efit

LiamPattinson committed:
Clean up and comments on GENE tracer_efit

LiamPattinson committed:
[skip ci] Apply black changes

LiamPattinson committed:
[skip ci] Apply black changes

LiamPattinson committed:
Remove overwrite_dpsidr parameter

LiamPattinson committed:
Remove overwrite_dpsidr parameter

LiamPattinson committed:
return new local geometry from normalise

Also removes normalisation side-effect from from_global_eq

LiamPattinson committed:
return new local geometry from normalise

Also removes normalisation side-effect from from_global_eq

LiamPattinson commented:
With the aim of promoting immutable behaviour, `normalise()` now returns a new object, and `from_global_eq` no longer updates the current normalisation as a side effect. In order to merge `LocalGeometry` and `FluxSurface`, it would be ideal to remove the `norms` argument from `from_global_eq` entirely, so I'll have a think about how best to achieve that. I also still have to fix a couple of broken tests and go over the docs.

LiamPattinson committed:
Improve Miller docs, move get_second_derivatives to LocalGeometry

LiamPattinson committed:
Improve Miller docs, move get_second_derivatives to LocalGeometry

LiamPattinson committed:
Improve MillerTurnbull docs

Some edits to Miller and LocalGeometry

LiamPattinson committed:
Improve MillerTurnbull docs

Some edits to Miller and LocalGeometry

LiamPattinson committed:
Improve MXH docs

Further improvements to other local geometry docs

LiamPattinson committed:
Improve MXH docs

Further improvements to other local geometry docs

LiamPattinson committed:
Improve Fourier GENE docs

LiamPattinson committed:
Improve Fourier GENE docs

LiamPattinson committed:
Improve Fourier CGYRO docs

LiamPattinson committed:
Improve Fourier CGYRO docs

LiamPattinson committed:
Can use from_global_eq without passing in norms

A step towards merging FluxSurface and LocalGeometry

LiamPattinson committed:
Can use from_global_eq without passing in norms

A step towards merging FluxSurface and LocalGeometry

LiamPattinson committed:
Add sympy code to generate Fourier GENE test_grad_r params

LiamPattinson committed:
Add sympy code to generate Fourier GENE test_grad_r params

LiamPattinson committed:
Fix broken fourier geometry tests

LiamPattinson committed:
Fix broken fourier geometry tests

LiamPattinson committed:
Remove defaults from most local geometry input parameters

LiamPattinson committed:
Remove defaults from most local geometry input parameters

LiamPattinson commented:
I think that's about as far as I'd like to take this PR. I'd still like to merge `FluxSurface` and `LocalGeometry`, but it's less essential. I've reworked the docs, removed most default `__init__` parameters, and fixed the broken tests. I also removed the `norms` argument from `from_global_eq`, as if I do get around to merging `FluxSurface` and `LocalGeometry`, it would be good to be able to do so without creating a `Normalisation` first.
